### PR TITLE
Scene Refactor

### PR DESCRIPTION
**Pull request by bagnell**
* Refactors the `Scene`'s private `executeCommands` function. Separates primitive updates, framebuffer updates/clears, draw command executions, and framebuffer resolves.
* Add ability to change the default viewport per pass. All commands will be executed in the default viewport unless overridden by the render state (e.g.  post-processing).

### CODE REVIEW

File: Source/Renderer/Context.js [hunk 1]
```diff
@@ -357,6 +357,9 @@ define([
         var ps = new PassState(this);
         var rs = RenderState.fromCache();
 
+        // default to the whole drawing buffer
+        ps.viewport = undefined;
```
Comment (pjcozzi) [inline Source/Renderer/Context.js]: Should this be the default in `PassState`?

File: Source/Scene/Scene.js [hunk 2]
```diff
@@ -526,6 +532,30 @@ define([
         this._cameraClone = Camera.clone(camera);
         this._screenSpaceCameraController = new ScreenSpaceCameraController(this);
 
+        // Keeps track of the state of a frame. FrameState is the state across
+        // the primitives of the scene. This state is for internally keeping track
+        // of celestial and environment effects that need to be updated/rendered in
+        // a certain order as well as updating/tracking framebuffer usage.
+        this._state = {
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: Perhaps name this `_environmentState`?

File: Source/Scene/Scene.js [hunk 3]
```diff
@@ -1444,92 +1424,31 @@ define([
...
         us.updateFrustum(frustum);
 
+        var skyBoxCommand = scene._state.skyBoxCommand;
         if (defined(skyBoxCommand)) {
             executeCommand(skyBoxCommand, scene, context, passState);
         }
 
-        if (defined(skyAtmosphereCommand) && skyAtmosphereVisible) {
-            executeCommand(skyAtmosphereCommand, scene, context, passState);
+        if (scene._state.isSkyAtmosphereVisible) {
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: This function could be cleaned up by assign `scene._state` to a local.  Same for a few other places in this file.

File: Source/Scene/Scene.js [hunk 4]
```diff
@@ -1684,6 +1586,24 @@ define([
     function updatePrimitives(scene) {
         var frameState = scene._frameState;
 
+        // Update celestial and terrestrial environment effects.
+        var renderPass = frameState.passes.render;
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: What do you think of moving this to a separate top-level function `updateEnvironment` like `updatePrimitives`?

**Comment by pjcozzi** [comment]
Just those minor comments.  Does this need new tests to have full coverage?

**Comment by bagnell** [comment]
>Does this need new tests to have full coverage?

It doesn't need any new tests. When there is a public API to change the viewport/camera/etc, new tests will be needed.

@pjcozzi This is ready.

**Comment by pjcozzi** [comment]
Looks good.

**Pull request merged**